src/tag_storage.rs

```diff
@@ -0,0 +1,214 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+///! This is the database that holds tags associated to various objects.
+///! It provides an api to manage Id <-> tags relationships.
+///! All users share the same tags for objects.
+
+use rusqlite::{ Connection, Result };
+use std::path::PathBuf;
+use util::{ Id, TagId };
+
+fn escape<T>(string: &Id<T>) -> String {
+    // http://www.sqlite.org/faq.html#q14
+    format!("{}", string).replace("'", "''")
+}
+
+/// Creates a unique key for a (id, tag) tuple.
+/// `SQlite` integers are i64 so we turn the hashed u64 into a String...
+fn create_key<T>(id: &Id<T>, tag: &Id<TagId>) -> String {
+    use std::hash::{ Hash, Hasher, SipHasher };
+
+    let mut hasher = SipHasher::new();
+    id.hash(&mut hasher);
+    tag.hash(&mut hasher);
+    format!("{}", hasher.finish())
+}
+
+pub struct TagStorage {
+    db: Connection,
+}
+
+impl TagStorage {
+    pub fn new(path: &PathBuf) -> Self {
+        println!("Opening db at {}", path.display());
+        let db = Connection::open(path).unwrap();
```

Comment:
I think that some of the issues above would disappear if you could open the connection lazily.